File: codes/core-service/app/schemas/views/workspace.py
from datetime import datetime

from pydantic import BaseModel, Field


class MemberPreviewOut(BaseModel):
    id: str
    display_name: str
    email: str


class WorkspaceMembersSummaryOut(BaseModel):
    total: int
    owner_count: int
    member_count: int
    owners_preview: list[MemberPreviewOut] = Field(default_factory=list)
    members_preview: list[MemberPreviewOut] = Field(default_factory=list)


class WorkspaceStatsViewOut(BaseModel):
    project_count: int
    total_task_count: int
    todo_count: int
    doing_count: int
    done_count: int
    archived_count: int
    high_priority_count: int
    health_percent: int | None


class WorkspaceProjectCardOut(BaseModel):
    id: str
    name: str
    description: str | None = None
    can_manage: bool
    todo_doing: int
    done_archived: int
    progress_percent: int


class WorkspaceDashboardOut(BaseModel):
    workspace_id: str
    name: str
    description: str | None = None
    created_at: datetime | None = None
    created_by_display_name: str | None = None
    can_edit_workspace: bool
    current_user_id: str
    members: WorkspaceMembersSummaryOut
    stats: WorkspaceStatsViewOut
    active_projects: list[WorkspaceProjectCardOut]
    total_active_projects: int


class DiscussionViewItemOut(BaseModel):
    id: str
    body: str
    created_at: datetime
    created_at_exact_label: str
    created_ago_label: str
    author_user_id: str
    author_display_name: str
    is_reply: bool
    completion_status: str
    is_author: bool
    project_id: str
    project_name: str
    item_id: str
    item_title: str


class WorkspaceDiscussionsViewOut(BaseModel):
    items: list[DiscussionViewItemOut]
    has_more: bool

File: codes/web/src/types/api/views/workspace.ts
export type MemberPreview = {
  id: string;
  display_name: string;
  email: string;
};

export type WorkspaceMembersSummary = {
  total: number;
  owner_count: number;
  member_count: number;
  owners_preview: MemberPreview[];
  members_preview: MemberPreview[];
};

export type WorkspaceStatsView = {
  project_count: number;
  total_task_count: number;
  todo_count: number;
  doing_count: number;
  done_count: number;
  archived_count: number;
  high_priority_count: number;
  health_percent: number | null;
};

export type WorkspaceProjectCard = {
  id: string;
  name: string;
  description?: string | null;
  can_manage: boolean;
  todo_doing: number;
  done_archived: number;
  progress_percent: number;
};

export type WorkspaceDashboardView = {
  workspace_id: string;
  name: string;
  description?: string | null;
  created_at?: string | null;
  created_by_display_name?: string | null;
  can_edit_workspace: boolean;
  current_user_id: string;
  members: WorkspaceMembersSummary;
  stats: WorkspaceStatsView;
  active_projects: WorkspaceProjectCard[];
  total_active_projects: number;
};

export type DiscussionViewItem = {
  id: string;
  body: string;
  created_at: string;
  created_at_exact_label: string;
  created_ago_label: string;
  author_user_id: string;
  author_display_name: string;
  is_reply: boolean;
  completion_status: string;
  is_author: boolean;
  project_id: string;
  project_name: string;
  item_id: string;
  item_title: string;
};

export type WorkspaceDiscussionsView = {
  items: DiscussionViewItem[];
  has_more: boolean;
};

export type WorkspaceCardView = {
  id: string;
  name: string;
  description?: string | null;
  project_count: number;
  todo_count: number;
  doing_count: number;
  done_count: number;
  archived_count: number;
  owners: Array<{ id: string; email: string; display_name: string; role: string; status: string }>;
  members: Array<{ id: string; email: string; display_name: string; role: string; status: string }>;
  my_workspace_role: string;
};
